<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Комментарий ревьюера</b>
    
Привет! Меня зовут Светлана Носова, я буду проверять твой проект. Если не против, давай общаться на «ты». Если тебе такой стиль общения не нравится, напиши об этом, будем на «вы». 

Сразу хочу обратить внимание на цветовую разметку.
Зеленым отмечены отличные решения и идеи, которые можно и нужно использовать в будущем. Желтым отмечено то, что можно или нужно оптимизировать, в зависимости от контекста. Красным обозначены пропущенные пункты задания и  ошибки, которые обязательно надо будет исправить, чтобы результаты были верные.
    
**Ответы на мои комментарии или доработки по моим рекомендациям попрошу выделить каким-нибудь цветом, кроме перечисленных.**
Также я попрошу тебя **не менять, не передвигать и не удалять мои комментарии**, чтобы при последующей проверке мне было проще ориентироваться.
    
Кроме того, комментарии у меня обозначены как заголовки. Это сбивает содержание, но при последующей проверке я могу снять с них заголовок, если тебе так будет удобнее. При этом они удобны, поскольку ты сразу можешь перейти к ним.   
        
<hr>
    
**По проекту:** в целом, работа выполнена хорошо, все оформлено понятно и аккуратно, есть анализ и промежуточные выводы. У меня остались некоторые вопросы, я все написала в комментариях. Там же я оставила и рекомендации по улучшению проекта. Буду ждать работу на повторное ревью.
    
<hr style="border-top: 3px solid purple; "></hr>
Комментарии можешь оставлять с помощью такого кода в Markdown ячейке:
    
    
    <div class="alert alert-info">
    <h2> Комментарий студента</h2>

    Тут текст.
    </div>
    
    
    <font color='red'> Этот код позволяет менять цвет текста. </font>     

<font color='red'> Выглядит это так. </font> 
  
    
Если не хочешь, чтобы твои комментарии отображались в содержании как заголовки, замени **h2** на **b** или добавь `<a class="tocSkip">` после фразы *Комментарий студента*.
</div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера</h2>
    
Правильно делаешь, что добавляешь краткое описание задачи. Это поможет быстро вспомнить, о чем проект, если ты к нему вернешься спустя какое-то время.
</div>

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
import pandas as pd
data = pd.read_csv('/content/drive/MyDrive/Inga/Data/data_project_1.csv')
print(data.head(20))
data.info()

    Unnamed: 0  children  ...   total_income                                 purpose
0            0         1  ...  253875.639453                           покупка жилья
1            1         1  ...  112080.014102                 приобретение автомобиля
2            2         0  ...  145885.952297                           покупка жилья
3            3         3  ...  267628.550329              дополнительное образование
4            4         0  ...  158616.077870                         сыграть свадьбу
5            5         0  ...  255763.565419                           покупка жилья
6            6         0  ...  240525.971920                       операции с жильем
7            7         0  ...  135823.934197                             образование
8            8         2  ...   95856.832424                   на проведение свадьбы
9            9         0  ...  144425.938277                 покупка жилья для семьи
10          10         2  ...  113943.491460                    п

**Вывод**

Импортировали библиотеку pandas. Методом .read_csv() загрузили файл и сохранили в переменную data.
При просмотре исходной таблицы видны отрицательные значения в продолжительности рабочего стажа (в days_employed), в столбце education есть значения, которые начинаются с заглавных букв (Среднее), total_income приведен до шестого знака после запятой. Заметны большие значения в days_employed и total_income. При просмотре информации, полученной методом .info(), замечены пропущенные значения в столбцах days_employed, total_income.   

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера</h2>
    
Хорошая вводная часть.
</div>

## Шаг 2. Предобработка данных

### Обработка пропусков

In [5]:
print(data.isnull().sum())
missing_data = data[data.isnull().any(axis=1)]
print(missing_data)

print(data['income_type'].unique())
print(missing_data['income_type'].unique())

data['days_employed'] = data['days_employed'].abs()



Unnamed: 0             0
children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64
       Unnamed: 0  children  ...  total_income                           purpose
12             12         0  ...           NaN                   сыграть свадьбу
26             26         0  ...           NaN                       образование
29             29         0  ...           NaN  строительство жилой недвижимости
41             41         0  ...           NaN  сделка с подержанным автомобилем
55             55         0  ...           NaN                   сыграть свадьбу
...           ...       ...  ...           ...                               ...
21489       21489         2  ...           NaN              сделка с автомобилем
21495       21

In [6]:
print(data['dob_years'].max())
print(data['dob_years'].min())

def age_groups(age):
    if age <20:
        return 'young'
    if 20 <= age < 55:
        return 'middle age'
    else:
        return 'senior'
    
data['age_group'] = data['dob_years'].apply(age_groups)    
    
data.loc[:,('total_income')] = data.groupby(['income_type', 'education', 'age_group'])['total_income'].apply(lambda x: x.fillna(x.median()))

data.loc[:,('days_employed')] = data.groupby('income_type')['days_employed'].apply(lambda x: x.fillna(x.median()))

print(data.isnull().sum())

data[data['total_income'].isnull()]

#remove these rows


75
0
Unnamed: 0          0
children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        4
purpose             0
age_group           0
dtype: int64


Unnamed: 0  children  ...                     purpose  age_group
3891         3891         1  ...  строительство недвижимости     senior
5936         5936         0  ...  покупка жилой недвижимости     senior
6670         6670         0  ...  покупка жилой недвижимости      young
18818       18818         0  ...           операции с жильем     senior

[4 rows x 14 columns]

In [12]:
data = data.dropna()

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера</h2>
    
Отлично. Раз уж у нас есть такая возможность, давай попробуем заменить пропуски медианой, но взятой по 2-3 характеристикам клиента? Группы можно собрать по разным параметрам. Например, тип занятости + образование + возрастная группа. Попробуешь реализовать это? Если что, я помогу. Столбец стажа можно так не заполнять, поскольку он нам не нужен. 
</div>
<div class="alert alert-info">
<h2> Комментарий студента</h2>

Добавлены изменения с учетом рекоммендаций:
1)    выявлены минимальный и максимальный возраст, чтобы убедиться в отсутствии выбросов:
    print(data['dob_years'].max())
    print(data['dob_years'].min())
    
2) Написана функция для сортировки по возрасту:
    def age_groups(age):
    if age <20:
        return 'young'
    if 20 <= age < 55:
        return 'middle age'
    else:
        return 'senior'
                      
3) Добавлен новый столбец с вызовом функции
                      data['age_group'] = data['dob_years'].apply(age_groups) 
                      
4) В группировку для 'total income' добавлены столбцы 'education' и 'age_group':
                      data.loc[:,('total_income')] = data.groupby(['income_type', 'education', 'age_group'])['total_income'].apply(lambda x: x.fillna(x.median()))
                       
</div>



**Вывод**

Используя функции .isnull().sum() видим, что в столбцах days_employed и total_income пропущено одинаковое количество значений - 2174. После отфильтровки строк с пропущенными значениями при помощи логической индексации, видим, что общее количество строк с пропущенными значениями 2174. Это означает что в каждой из таких сторок пропущены сразу два значения: days_employed и total_income. Это выглядим логичным, так как люди без работы не имеют дохода. Проверяем эту гипотезу. 

Вызовом функции unique() сморим, как обозначены люди без дохода и видим, что есть отдельная категория 'безработный'. При проверке уникальных значений для таблицы с пропущенными значениями missing_data, видим, что безработные туда не входят. Значит, значения пропущены случайным образом. 

Все пропуски помечены NaN. Их необходимо заполнить медианными данными для соответствующих категорий занятости ('пенсионер' 'госслужащий' 'компаньон' 'сотрудник' 'предприниматель'). Перед рассчетом медианы, избавимся от отрицательныз значений в столбце 'days_employed', сохранив их как положительные.  

Заполним пропуски на медианные значения для соответствующих категорий используя метод apply(). Затем проверим заполнение вызовом функции isnull().sum(). Пропущенных значений больше нет.  

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера</h2>
    
Да, очень похоже на то, что пропуски случайны. Скорее всего имеем склейку разных таблиц.</div>

### Замена типа данных

In [13]:
data['days_employed'] = data['days_employed'].astype(int)
data['total_income'] = data['total_income'].astype(int)
       

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Вывод

Для удобства работы с данными, заменим тип данных столбцов days_employed и total_income с float на integer испольщуя метод .astype(int). Десятичные цифры в данном случае не важны, их можно опустить для ускорения времени обработки.


### Обработка дубликатов

In [14]:
from collections import Counter
print(Counter(data['education']))
data['education'] = data['education'].str.lower()
print(Counter(data['education']))

print(Counter(data['family_status']))
data['family_status'] = data['family_status'].str.lower()
print(Counter(data['family_status']))
print(Counter(data['income_type']))
print(Counter(data['purpose']))
print(Counter(data['gender']))

print(data.duplicated().sum()) 
print(data.shape)

data.drop_duplicates(keep = 'first', inplace = True)

print(data.duplicated().sum()) 
print(data.shape)


Counter({'среднее': 13750, 'высшее': 4717, 'СРЕДНЕЕ': 772, 'Среднее': 711, 'неоконченное высшее': 668, 'ВЫСШЕЕ': 274, 'Высшее': 267, 'начальное': 250, 'Неоконченное высшее': 47, 'НЕОКОНЧЕННОЕ ВЫСШЕЕ': 27, 'НАЧАЛЬНОЕ': 17, 'Начальное': 15, 'ученая степень': 4, 'Ученая степень': 1, 'УЧЕНАЯ СТЕПЕНЬ': 1})
Counter({'среднее': 15233, 'высшее': 5258, 'неоконченное высшее': 742, 'начальное': 282, 'ученая степень': 6})
Counter({'женат / замужем': 12378, 'гражданский брак': 4177, 'Не женат / не замужем': 2813, 'в разводе': 1193, 'вдовец / вдова': 960})
Counter({'женат / замужем': 12378, 'гражданский брак': 4177, 'не женат / не замужем': 2813, 'в разводе': 1193, 'вдовец / вдова': 960})
Counter({'сотрудник': 11117, 'компаньон': 5085, 'пенсионер': 3855, 'госслужащий': 1459, 'безработный': 2, 'студент': 1, 'предприниматель': 1, 'в декрете': 1})
Counter({'свадьба': 797, 'на проведение свадьбы': 777, 'сыграть свадьбу': 774, 'операции с недвижимостью': 676, 'покупка коммерческой недвижимости': 664, 'по

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https:

In [15]:
#Для обработки дубликатов в столбцах 'education_id'/'education' и 'family_status_id'/'family_status' были созданы словари:
    
education_dict = data[['education_id','education']]
family_status_dict = data[['family_status_id', 'family_status']]

# Затем было проведено удаление дубликатов из словарей припомощи метода drop_duplicates().Результат был проверен выводом на экран отсортированных значений.

education_dict = education_dict.drop_duplicates().reset_index(drop=True)
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)
print(education_dict.sort_values(by = 'education_id'))

print(family_status_dict.sort_values(by = 'family_status_id'))

   education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень
   family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера</h2>
    
Поскольку у нас нет уникальных идентификаторов пользователей, эти дубликаты вполне могут быть разными людьми. А вот дубликаты записей в столбцах вроде **Среднее** образование и **среднее** образование удалить надо было (путем приведения к одному регистру), это верно. </div>




**Вывод**

1) Методом Counter проверим количества уникальных значений для столбцов типа object. Видим, что в столбцах 'education' и 'family_status' много дубликатов из-за использования заглавных букв. Приведем содержимое этих двух столбцов к единому формату используя функцию str.lower(). Проверим результат методом Counter.
2) Вызовом функции duplicated() находим общее количество дупликатов в таблице. Методом drop_duplicates удаляем полные дубликаты в таблице, оставляя первую строку при помощи keep = 'first' и избавляясь только от дублирующей строки. Проверяем результат методом shape до и после удаления. Видим, что удалены только 71 дублирующих строки. Полные дубликаты могли появиться при объединении нескольких меньших датасетов в один большой.


<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid">
<h2> Комментарий ревьюера </h2>
    
В задаче нас просят обратить внимание на словари. Словарем можно назвать пару столбцов `education` и `education_id`, поскольку они взаимозаменяемы и из них можно слепить словарь (структуру данных dict). Есть еще что-нибудь подобное в таблице?
    
<div class="alert alert-info">
<h2> Комментарий студента</h2>

Для обработки дубликатов в столбцах 'education_id'/'education' и 'family_status_id'/'family_status' были созданы словари:
    
education_dict = data[['education_id','education']]
family_status_dict = data[['family_status_id', 'family_status']]

Затем было проведено удаление дубликатов из словарей припомощи метода drop_duplicates().Результат был проверен выводом на экран отсортированных значений.

education_dict = education_dict.drop_duplicates().reset_index(drop=True)
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)
print(education_dict.sort_values(by = 'education_id'))

print(family_status_dict.sort_values(by = 'family_status_id')) 
</div>
</div>

### Лемматизация

In [16]:
from pymystem3 import Mystem
m = Mystem()
lemmas = [[m.lemmatize(word) for word in data['purpose']]]  
#print(lemmas)
import numpy as np
n_lemmas = np.array(lemmas)
df_lemmas = pd.Series(n_lemmas.reshape(-1))

data['df_lemmas'] = df_lemmas

data.head()

Installing mystem to /root/.local/bin/mystem from http://download.cdn.yandex.net/mystem/mystem-3.1-linux-64bit.tar.gz


KeyboardInterrupt: ignored

# Начало комментария

In [ ]:
# Сейчас будет код, который Вы можете пропустить, т.к. у Вас уже есть столбец с леммами, полученный с помощью mystem

In [24]:
# СТАРТ 
data_with_lemmas = pd.read_csv('/content/drive/MyDrive/Inga/Data/new_data')

In [19]:
!pip install pymorphy2

     |████████████████████████████████| 61kB 5.3MB/s 
     |████████████████████████████████| 8.2MB 14.4MB/s 


In [27]:
data_with_lemmas # отсюда нужен только столбец с названием 0, остальные данные тут НЕОБРАБОТАННЫЕ - это СТАРАЯ версия данных.

Unnamed: 0  ...                                                  0
0               0  ...                    ['покупка', ' ', 'жилье', '\n']
1               1  ...          ['приобретение', ' ', 'автомобиль', '\n']
2               2  ...                    ['покупка', ' ', 'жилье', '\n']
3               3  ...       ['дополнительный', ' ', 'образование', '\n']
4               4  ...                  ['сыграть', ' ', 'свадьба', '\n']
...           ...  ...                                                ...
21520       21520  ...         ['операция', ' ', 'с', ' ', 'жилье', '\n']
21521       21521  ...      ['сделка', ' ', 'с', ' ', 'автомобиль', '\n']
21522       21522  ...                             ['недвижимость', '\n']
21523       21523  ...  ['на', ' ', 'покупка', ' ', 'свой', ' ', 'авто...
21524       21524  ...    ['на', ' ', 'покупка', ' ', 'автомобиль', '\n']

[21525 rows x 16 columns]

In [ ]:
# ФИНИШ

In [20]:

import pymorphy2
morph = pymorphy2.MorphAnalyzer()

In [21]:
# функция для выбора только существительных из списка строк
def only_nouns(list_of_strings):
    buf_list = []
    for word in list_of_strings:
        #print(word)
        razbor = morph.parse(word)[0]
        if 'NOUN' in razbor.tag:
            buf_list.append(word)
    
    return buf_list # возвращает список строк


In [41]:
lem = pd.Series(data_with_lemmas.iloc[:, 15], name='df_Lemmas') # вырезаю из data_with_lemmas столбец с собственно леммами.
#В Вашем коде этот столбец лежит здесь: data['df_lemmas'], поэтому Вы можете пропустить блок кода, который обозначен СТАРТ и ФИНИШ. Просто передайте его сюда через lem = data['df_lemmas']

In [42]:
#lem[0] проверочная ячейка

"['покупка', ' ', 'жилье', '\\n']"

In [46]:
# обращаем внимание: сейчас элементы lem - СТРОКИ!

In [44]:
lem = lem.apply(lambda x : x.replace("[", "").replace("]", "").replace("\'", "").replace(",", "").replace("\\n", "").replace("  ", '').split()) # эту ячейку запускаем РОВНО один раз

In [47]:
# обращаем внимание опять: теперь элементы lem - СПИСКИ!

In [46]:
# к текущему моменту в столбце lem остались списки строк со словами:

lem[4]

['свадьба']

In [ ]:
# обратите внимание, что если сейчас еще раз запустить код двумя ячейками выше, то выпадет ошибка:
# раньше lem содержал в себе СТРОКИ а сейчас там СПИСКИ - никаких методов .replace() у списков нет!

In [45]:
lem = lem.apply(only_nouns) # эту ячейку тоже один раз

In [48]:
lem

0                  [покупка, жилье]
1        [приобретение, автомобиль]
2                  [покупка, жилье]
3                     [образование]
4                         [свадьба]
                    ...            
21520             [операция, жилье]
21521          [сделка, автомобиль]
21522                [недвижимость]
21523         [покупка, автомобиль]
21524         [покупка, автомобиль]
Name: df_Lemmas, Length: 21525, dtype: object

In [50]:
# осталось оставить только последние слова в этих списках - и далее оформлять результат - 
# подсматривайте в код last_version_updated_july_5 и старайтесь следовать рекомендациям по оформлению кода - как видите, они помогают избегать ошибок!

# Конец комментария

**Вывод**

Лемматизируем строки в столбце 'purpose'. Для этого выделим основу каждого слова word в столбце и сохраним как переменную lemmas.



### Категоризация данных

In [ ]:
#print(data['purpose'].value_counts())

#def category_purpose(purpose):
#    for i in lemmas:
#        if 'жилье' in i:
#            return 'операции с недвижимостью'
#        if 'недвижимость' in i:
#            return 'операции с недвижимостью'
#        if 'автомобиль' in i:
#            return 'операции с автомобилем'
#        if 'свадьба' in i:
#            return 'свадьба'  
#        if 'образование' in i:
#            return 'образование'     
#    






<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера</h2>
    
Согласна с категориями. Но если у тебя будет миллион строк и 100 категорий? Проще ведь сначала лемматизировать, а потом тот же Counter применить.  </div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера</h2>
    
Не забудь исправить ошибку. </div>

**Вывод**

Методом value_counts проверяем какие значения встречаются в столбце 'purpose'. Видно, что они и те же цели написаны по-разному (например, 'свадьба': 797, 'на проведение свадьбы': 777, 'сыграть свадьбу' и тд.). В функции создадим сортировку по категорям для целей кредита и перезапишем столбец purpose используя категории. Методом Counter проверим новые значения столбца purpose.  

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [ ]:
# Проверим количество детей на наличие артефактов:
print(data['children'].value_counts())

# Видим, что 47 строк имеют отрицательное количество детей (-1). 
#Скорее всего, знак минус был добавлен ошибочно. Избавимся от отрицательных значений методом abs:
data['children'] = data['children'].abs()

# 76 строк имеют количестыо делей равное 20. Учитывая что максимальное значение детей 5 (если отбросить 20),
#говорит о том, что данные наиболее вероятно были введены с ошибкой. Перезапишем 20 детей как 2:

def change_amount_of_children(x):
    if x == 20:
        return 2
    else:
        return x

  
data['children'] = data['children'].apply(change_amount_of_children)

# Проверим результат методом value_counts()
print(data['children'].value_counts())

In [ ]:
debt_by_children_group = data.groupby('children')['debt'].sum()
applicants_by_children_group = data.groupby('children')['children'].count() 

print(debt_by_children_group)
print(applicants_by_children_group)

perc_of_debt_by_children = (debt_by_children_group / applicants_by_children_group) * 100

print((perc_of_debt_by_children).sort_values())

**Вывод**

Коэффициент корреляции между наличием детей и задолженностей по кредиту равен 0.03. Явной зависимости между этими двумя параметрами нет.

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера</h2>
    
Это коэфф-т линейной связи. А если связь нелинейная? Выведи проценты невозврата. Для этого надо число долгов разделить на общее число кредитов внутри каждой группы. Ниже аналогично. </div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Подсчитаем количество долга в зависимости от наличния и количества детей. Для этого сгруппируем данные по столбцу 'children' и подсчитаем количество долгов по сгруппированным данным:    
debt_by_children_group = data.groupby('children')['debt'].sum()
    
Затем подсчитаем общее количество подавшихся на кредит также сгруппированных по наличию и количеству детей:    
applicants_by_children_group = data.groupby('children')['children'].count() 

Распечатаем предварительные данные:
print(debt_by_children_group)
print(applicants_by_children_group)

Вылислим процент невозврата по сгруппированным данным и выведем данные на экран:    
perc_of_debt_by_children = (debt_by_children_group / applicants_by_children_group) * 100

print(perc_of_debt_by_children)
</div>

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера</h2>
    
20 детей это не ошибка? </div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Проверим количество детей на наличие артефактов:
print(data['children'].value_counts())

Видим, что 47 строк имеют отрицательное количество детей (-1). 
Скорее всего, знак минус был добавлен ошибочно. Избавимся от отрицательных значений методом abs:
data['children'] = data['children'].abs()

76 строк имеют количестыо делей равное 20. Учитывая что максимальное значение детей 5 (если отбросить 20), это
говорит о том, что данные наиболее вероятно были введены с ошибкой. Перезапишем 20 детей как 2:

def change_amount_of_children(x):
    if x == 20:
        return 2
    else:
        return x

  
data['children'] = data['children'].apply(change_amount_of_children)

Проверим результат методом value_counts()
print(data['children'].value_counts())
</div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Вывод:
Самый низкий процент невозврата наблюдается среди людей с 5 детьми. Однако это неверный вывод, так как таких людей всего 9. соответственно, данные по этой группе не могут быть использованы для исчерпывающего анализа. 
Следующий самый низкий процент невозврата (7.51%) наблюдается среди людей без детей. Самый высокий процент наблюдается для людей с 4 детьми (9.75%), однако таких людей всего 41, что значительно меньше в сравнении с количеством людей без детей (14149). В целом с увеличением количества детей процент невозврата растет. 
    
</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [ ]:
print(data['family_status'].value_counts())

debt_by_family_status_group = data.groupby('family_status')['debt'].sum()
applicants_by_family_status_group = data.groupby('family_status')['family_status'].count() 

perc_of_debt_by_family_status = (debt_by_family_status_group / applicants_by_family_status_group) * 100

print((perc_of_debt_by_family_status).sort_values())

**Вывод**

Коэффициент корреляции между семейным положением и задолженностей по кредиту равен -0.008. Явной зависимости между этими двумя параметрами нет.
<div class="alert alert-info">
<h2> Комментарий студента</h2>

Вывод:
Самый низкий процент невозврата наблюдается для людей со статусом 'вдовец / вдова', однако общее количество таких людей в 10 раз меньше по сравнению с людьми в статусе 'женат / замужем'. 
Самый высокий процент наблюдался для людей со статусом 'Не женат / не замужем' (9.74%).    
    
</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [ ]:
data['debt'].corr(data['total_income']) # линейной корреляции нет

# Разобьем данные по доходу на квартили, чтобы получить рамки доходом для группировки по группам: 
print(data['total_income'].quantile([0.25, 0.50, 0.75], interpolation='nearest'))

# Группируем используя полученные квартили:
def income_levels(x):
    if x < 107702:
        return 'low_income'
    if 107702 <= x < 143302:
        return 'middle'
    if 143302 <= x < 198149:
        return 'middle_high'
    if x > 198149:
        return 'high'
    
data['income_group'] = data['total_income'].apply(income_levels)   

print(data['income_group'].value_counts())

debt_by_income_group = data.groupby('income_group')['debt'].sum()
applicants_by_income_group = data.groupby('income_group')['income_group'].count() 

perc_of_debt_by_income = (debt_by_income_group / applicants_by_income_group) * 100

print((perc_of_debt_by_income).sort_values())

**Вывод**

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Линейной корреляции между уровнем дохода и возвратом кредита в срок выявлено не было. 
    
После распреденения данных по уровню дохода с использованием квартилей и расчета доли невыплат по кредиту для сгруппированных данных, видим следующее: самый низкий процент невозврата (6.96%) наблюдается для группы с высоким уровнем дохода (выше 198149). Самый высокий процент (8.88%) наблюдается для для группы с уровнем дохода 107702 - 143302.
</div>

- Как разные цели кредита влияют на его возврат в срок?

In [ ]:
data['purpose']=data['purpose'].astype('category').cat.codes
data['debt'].corr(data['purpose'])


**Вывод**

Для анализа корреляции нужно привести значения в столбце purpose к категориальному типу. При использовании метода corr, коэффициент корреляции 0.001. Значительной корреляции между величинами нет.

## Шаг 4. Общий вывод

Анализ показал, что ни один из перечисленных факторов не оказывает значительного влияния на возврат кредита в срок. 

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid">
<h2> Комментарий ревьюера</h2>
    
После корректировок не забудь привести проценты невозврата в срок у самых надежных и самых безответственных клиентов для каждой группы. Например: самые надежные клиенты это люди, берущие кредит на операции с недвижимостью (такой-то процент невозврата в срок), а самые безответственные те, кто обращается в банк с целью получить деньги на покупку автомобиля (такой-то процент). И так для каждый группы. Это хороший тон, — подкреплять вывод полученными ранее результатами. Кроме того, так ты не оставляешь заказчику шансов неверно интерпретировать вывод.
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [ ]  объяснено, по какому принципу заполнены пропуски;
- [ ]  заменен вещественный тип данных на целочисленный;
- [ ]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [ ]  удалены дубликаты;
- [ ]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [ ]  описаны возможные причины появления дубликатов в данных;
- [ ]  выделены леммы в значениях столбца с целями получения кредита;
- [ ]  описан процесс лемматизации;
- [ ]  данные категоризированы;
- [ ]  есть объяснение принципа категоризации данных;
- [ ]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [ ]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [ ]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [ ]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [ ]  в каждом этапе есть выводы;
- [ ]  есть общий вывод.